### PR TITLE
Avoid one unnecessary memory allocation in XNNPACK integration.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35896 [WIP] Add XNNPACK Clamp.
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* #35354 Support for XNNPACK max pooling operator.
* #35351 Propagate input tensor names in XNNPACK backend.
* **#35350 Avoid one unnecessary memory allocation in XNNPACK integration.**

Currently we call input.contiguous() on the input tensor resulting in an
unecessary allocation and copy in cases where the input is not contiguous
with regards to the requested memory format.  The reason is that in such
scenarios, this call re-allocates and copies the input tensor into
contiguous storage, only for this newly allocated tensor to be used as
the source of another copy to the final destination.  Instead, if we copy
into the destination directly in such circumstances, we will save an
allocation and a copy.

Differential Revision: [D20656798](https://our.internmc.facebook.com/intern/diff/D20656798)

### CODE REVIEW

**Comment by AshkanAliabadi** [comment]
Do we have any tests that I can use to validate that the results are correct?  I'm not sure this code is exercised through CI.

**Comment by kimishpatel** [comment]
> Do we have any tests that I can use to validate that the results are correct? I'm not sure this code is exercised through CI.

You can try `python test/test_xnnpack_integration.py`. I suggest add, in one or two tests, trying different memory formats for inputs, including channels first and channels last.

File: aten/src/ATen/native/xnnpack/Factory.cpp [hunk 1]
```diff
@@ -35,17 +34,31 @@ Tensor empty_with_tail_padding(
...
+Tensor allocate_padded_contiguous_if_needed(
+    const Tensor& input,
+    const c10::MemoryFormat memory_format) {
+  const auto* const allocator = input.storage().allocator();
+  const auto* const guarding_allocator = get_guarding_allocator();
+
+  // If the allocators are the same and the memory is contiguous in the requested
+  // format, then there is no need to reallocate the tensor.
+
+  if ((allocator == guarding_allocator) && input.is_contiguous(memory_format)) {
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/Factory.cpp]: This function is called only with `MemoryFormat::ChannelsLast`, right? If there is no other use case, then I am wondering if we should make that explicit in the function name and not pass memory_format argument? Just a thought.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
LG for the most part. Can you add the test I mentioned in the comment or try the existing tests?

**Review by dreiss — APPROVED** [review]

**Pull request closed without merging**